### PR TITLE
GEODE-8773: Test that PING and CLOSE_CONNECTION are not implicitly transactional.

### PR DESCRIPTION
**Pull request by mreddington**

### CODE REVIEW

**Comment by pdxcodemonkey** [comment]
@dihardman @karensmolermiller @davebarnes97

File: cppcache/integration/framework/Cluster.cpp [hunk 1]
```diff
@@ -110,6 +109,12 @@ void Locator::start() {
                      .withPreferIPv6(cluster_.getUseIPv6())
                      .withJmxManagerStart(true);
 
+  if(cluster_.getLogLevel().empty()) {
```
Comment (pivotal-jbarrett) [inline cppcache/integration/framework/Cluster.cpp:112]: Why pull this out of the `.withLogLevel` model? Why not have cluster default non-empty value and then always pass cluster value through to `gfsh.withLogLevel()`?

File: cppcache/integration/test/TransactionsTest.cpp [hunk 2]
```diff
@@ -36,6 +40,30 @@ using apache::geode::client::Pool;
 using apache::geode::client::Region;
 using apache::geode::client::RegionShortcut;
 
+std::string getServerLogName() {
```
Comment (pivotal-jbarrett) [inline cppcache/integration/test/TransactionsTest.cpp:43]: Using logs to validate tests is really dangerous. Log messages change and aren't part of API. Is there a more concrete way to test this? I would think a unit tests that asserts these two methods set the correct 'no tx' value of `-1` should be sufficient.

File: cppcache/src/TcrMessage.cpp [hunk 3]
```diff
@@ -2096,14 +2096,9 @@ TcrMessagePing::TcrMessagePing(DataOutput* dataOutput, bool decodeAll) {
   m_request->writeInt(static_cast<int32_t>(
       0));  // 17 is fixed message len ...  PING only has a header.
   m_request->writeInt(static_cast<int32_t>(0));  // Number of parts.
-  // int32_t txId = TcrMessage::m_transactionId++;
-  // Setting the txId to 0 for all ping message as it is not being used on the
-  // SERVER side or the
-  // client side.
-  m_request->writeInt(static_cast<int32_t>(0));
+  m_request->writeInt(m_txId);
```
Comment (pivotal-jbarrett) [inline cppcache/src/TcrMessage.cpp:2099]: A ping is not a transactional item. It should send `-1`.

File: cppcache/src/TcrMessage.cpp [hunk 4]
```diff
@@ -2114,8 +2109,7 @@ TcrMessageCloseConnection::TcrMessageCloseConnection(DataOutput* dataOutput,
   m_request->writeInt(m_msgType);
   m_request->writeInt(static_cast<int32_t>(6));
   m_request->writeInt(static_cast<int32_t>(1));  // Number of parts.
-  // int32_t txId = TcrMessage::m_transactionId++;
-  m_request->writeInt(static_cast<int32_t>(0));
+  m_request->writeInt(m_txId);
```
Comment (pivotal-jbarrett) [inline cppcache/src/TcrMessage.cpp:2112]: Not transactional, should send `-1`.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]
I take all my previous comments back on the grounds this change should not be made at all. To remain consistent, however odd, with the Java client these messages should continue to set transaction ID to `0`. The fix should be made in the Geode server sources at make BaseCommand.shouldMasqueradeForTx consistent with all other checks against `NOTX`. In all other checks explicit equality is checked, not for greater or less than.

Change
```java
  protected boolean shouldMasqueradeForTx(Message clientMessage,
      ServerConnection serverConnection) {
    return serverConnection.getClientVersion().isNotOlderThan(KnownVersion.GFE_66)
        && clientMessage.getTransactionId() > TXManagerImpl.NOTX;
  }
```
to
```java
  protected boolean shouldMasqueradeForTx(Message clientMessage,
      ServerConnection serverConnection) {
    return serverConnection.getClientVersion().isNotOlderThan(KnownVersion.GFE_66)
        && clientMessage.getTransactionId() != TXManagerImpl.NOTX;
  }
```

**Comment by pivotal-jbarrett** [comment]
> The fix should be made in the Geode server sources at make BaseCommand.shouldMasqueradeForTx consistent with all other checks against `NOTX`. In all other checks explicit equality is checked, not for greater or less than.

Pointed out offline to me, this won't fix the original issue either since `0 != NOTX` is `true` too. 🤦 

I really think we should understand why `0` was used here and on both the C++ and Java versions of these messages.

File: cppcache/integration/framework/Cluster.cpp (hunk 1 above)
Comment (pdxcodemonkey) [reply cppcache/integration/framework/Cluster.cpp]: Switched to a simple unit test, framework changes are gone.

File: cppcache/integration/test/TransactionsTest.cpp (hunk 2 above)
Comment (pdxcodemonkey) [reply cppcache/integration/test/TransactionsTest.cpp]: Done.

**Comment by pdxcodemonkey** [comment]
Per the larger team, Java side is not using 0 for these, only for certain server-initiated messages.  Java client sets NOTX for PING and CLOSE_CONNECTION, native client should as well.

File: cppcache/src/TcrMessage.cpp [hunk 5]
```diff
@@ -2090,17 +2090,16 @@ TcrMessageReply::TcrMessageReply(bool decodeAll,
 
 TcrMessagePing::TcrMessagePing(DataOutput* dataOutput, bool decodeAll) {
   m_msgType = TcrMessage::PING;
+  // -1 is NOTX constant server-side, PING should *never* be part of a
+  // transaction
+  m_txId = -1;
```
Comment (pivotal-jbarrett) [inline cppcache/src/TcrMessage.cpp:2095]: Isn't this the default? Look at `TcrMessage` default constructor.

File: cppcache/src/TcrMessage.cpp [hunk 6]
```diff
@@ -2090,17 +2090,16 @@ TcrMessageReply::TcrMessageReply(bool decodeAll,
...
   m_request->writeInt(m_msgType);
   m_request->writeInt(static_cast<int32_t>(
       0));  // 17 is fixed message len ...  PING only has a header.
   m_request->writeInt(static_cast<int32_t>(0));  // Number of parts.
-  // int32_t txId = TcrMessage::m_transactionId++;
-  // Setting the txId to 0 for all ping message as it is not being used on the
-  // SERVER side or the
-  // client side.
-  m_request->writeInt(static_cast<int32_t>(0));
+  m_request->writeInt(m_txId);
```
Comment (pivotal-jbarrett) [inline cppcache/src/TcrMessage.cpp:2102]: Look at `TcrMessage::writeHeader` and see that almost all the messages are using this method to do some of the message bits in here. Perhaps it makes sense to normalize this message to match.

File: cppcache/src/TcrMessage.cpp [hunk 7]
```diff
@@ -2109,13 +2108,16 @@ TcrMessagePing::TcrMessagePing(DataOutput* dataOutput, bool decodeAll) {
...
+  // of a transaction
+  m_txId = -1;
   m_decodeAll = decodeAll;
   m_request.reset(dataOutput);
   m_request->writeInt(m_msgType);
   m_request->writeInt(static_cast<int32_t>(6));
   m_request->writeInt(static_cast<int32_t>(1));  // Number of parts.
   // int32_t txId = TcrMessage::m_transactionId++;
-  m_request->writeInt(static_cast<int32_t>(0));
+  m_request->writeInt(m_txId);
```
Comment (pivotal-jbarrett) [inline cppcache/src/TcrMessage.cpp:2120]: See comment about this in ping message.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

**Comment by pivotal-jbarrett** [comment]
Should we also have a `NOTX` constant like the Java client?

**Review by pivotal-jbarrett — APPROVED** [review]

**Review by pivotal-jbarrett — APPROVED** [review]

**Pull request merged**